### PR TITLE
[Bugfix][Kernel] Fix top-k selection when the radix threshold bin overflows the smem stash

### PR DESCRIPTION
**Pull request by Dovis01**
# Summary

## Purpose

The radix-select top-k paths used by the DSA sparse-attention indexers (`persistent_topk` / `cooperative_topk` / `top_k_per_row`) stash the stage-1 coarse threshold bin in shared memory before refinement. When that bin holds
more candidates than the stash can take, candidates beyond capacity were silently dropped in atomic arrival order — they never reached refinement, so the selected set could differ from `torch.topk` on rows with tied or tightly clustered scores. Related latent issues on the same paths: selection scalars were only initialized inside the threshold-finder branch (stale shared-memory consumption across rows of the persistent loop), and stash/selection stores were unguarded.

Fixed in all four selection paths:

| path | stash |
|---|---|
| `histogram_2048_topk` (decode, seq_len <= 8192) | 3708 |
| `histogram_256_topk` (medium, seq_len <= 32768) | 4096 |
| `FilteredTopKUnifiedKernel` (batch > 32) | 16384 |
| `histogram_4096_topk` (short path, tie buffer) | TopK |

* An oversized threshold bin now descends the remaining FP32 key bytes (24 -> 16 -> 8 -> 0) until the participant set fits the stash; a capacity clip only happens after full-key equality, where the remaining candidates are exact ties and any subset is a valid selection.
* Selection-state scalars are reset unconditionally, before the finder; a no-finder round degenerates to a bounded path instead of consuming stale state.
* Every stash and selection store is guarded by its capacity; an underfilled selection slot emits `-1` instead of a stale index.
* `histogram_2048_topk` published the threshold-bin population by reading `histo[threshold]` — which aliases `bufs[0]` (`histo = decode_smem`, `bufs[0] = decode_smem + 768`) in the same barrier interval as the collection loop's writes. The finder thread now publishes the count from its register into the unused scalar slot `sPOP`.

Trigger condition (before the fix): coarse-bin population > stash, i.e. strongly tied scores (e.g. relu-weighted fp8 dot products, quantized/pooled logits). Not reached by every deployment — measured on GLM-5.3-Flash indexer
logits — but any DSV3.2/V4 sparse-MLA, Qwen4-exp QSA, or GLM kpool workload with heavy ties can hit it.

Concept originates from the SGLang fix for the same class of bug [sglang#37625](https://github.com/sgl-project/sglang/pull/37625); this is an independent port to vLLM's kernels (the Xid 31 shared-memory overrun reported there does not reproduce in vLLM — vLLM's stash stores were already capacity-guarded; the selection divergence was).

## Test Plan

```bash
# Regression suite incl. new oversized-bin cases (tight clusters,
# clusters separating at each key byte, all-equal rows, state reuse across
# rows, per-kernel-path coverage), exact value comparison vs torch.topk:
pytest tests/kernels/test_top_k_per_row.py -q

# Differential harness: rows 6/36 x lengths 700..80000 x 6 score
# distributions (gaussian, relu-dot, pooled-relu, quantized, tight cluster,
# all-equal), value-multiset comparison vs torch.topk at tolerance 0:
python <path to harness>/diff_topk.py 512   # 2048  and 4096
```

## Test Result

* `pytest tests/kernels/test_top_k_per_row.py -q`: **189 passed, 19 skipped** (H100, B200 CUDA 13.0).
* Differential harness: **78/78 shapes match `torch.topk` exactly** (tolerance 0) on post-fix builds — the fix changes selection only on rows whose threshold bin overflows, which were wrong before; realistic distributions are bit-identical.
* No model-eval delta expected on GLM-5.3-Flash serving; GPQA/AIME scores outputs are unchanged by this patch.

### CODE REVIEW

**Comment by jschmied** [comment]
@Dovis01 We are fixing the same defect from opposite ends and had not found each other — #55122 has
been open on this kernel since before this PR, and #53287 is a third. Cross-linking so all three do
not stall in parallel, and because I think they are more complementary than competing.

**Where we agree.** Same root cause: candidates past stash capacity dropped in arrival order, so the
selected *set* diverges from `torch.topk` on tied or tightly clustered rows. Your descend-the-key-bytes
approach fixes that while keeping the buffers; ours removes the buffers and rescans per key byte. Both
give an exact set.

**Where they differ, and it is not a style question.** This PR keeps `atomicAdd` for the output slot:

```
const int pos        = atomicAdd(&decode_smem[sOUT_abs], 1);
const int output_pos = atomicAdd(&shared_output_count, 1);
```

so the *order* of the emitted indices still depends on thread arrival. That matters here because the
sparse attention sums the selected keys in output order, so an order-only change still forks greedy
decoding between identical requests — which is the bug #54521 reports and what #55122 exists to fix.
Your clip "after full-key equality, where any subset is a valid selection" is sound for exactness and
not sufficient for reproducibility: *which* valid subset you get can still vary run to run.

Two checkable consequences, offered as a suggestion rather than a criticism:

1. Your three new tests are all `..._oversized_threshold_bin` — they compare against `torch.topk`
   once. **Calling the same kernel 6× on one input and requiring bit-identical output** would show
   this directly; on our GB10 the unmodified kernel reproduces its own output on **0 of 56** shapes.
2. If you want the order property, the cheap version is a packed `BlockScan` over the (greater, equal)
   flags instead of the atomic — it costs one scan and removes the arrival dependence entirely.

**What you cover that we deliberately do not.** `cooperative_topk` and `topk_histogram_4096`. #55122
bypasses the 2048-bin path rather than fixing it, and `cooperative_topk` we cannot even exercise —
all 51 of its cases fail on sm_121 with `cooperative_topk launch failed: invalid argument`
(cluster launch rejected on GB10). Your PR is the only one touching those.

Useful pointer from your side that I had missed: the SGLang origin
([sglang#37625](https://github.com/sgl-project/sglang/pull/37625)) and that vLLM's stash stores were
already capacity-guarded, so only the selection divergence was live here. That is a cleaner statement
of the scope than mine.

Happy to run your branch on a GB10 (sm_121) against our determinism harness — 56 shapes, 6 calls each,
exact-reference comparison — if that would help. It would answer the set question and the order
question in one pass.
